swap/src/protocol/alice/spot_price.rs

```diff
@@ -197,3 +207,43 @@ impl From<OutEvent> for alice::OutEvent {
         }
     }
 }
+
+impl<LR> From<Error<LR>> for spot_price::Error
+where
+    LR: LatestRate + Debug,
+{
+    fn from(error: Error<LR>) -> Self {
+        match error {
+            Error::ResumeOnlyMode => spot_price::Error::NoSwapsAccepted,
+            Error::MaxBuyAmountExceeded { max, buy } => {
+                spot_price::Error::MaxBuyAmountExceeded { max, buy }
+            }
+            Error::BalanceTooLow { buy } => spot_price::Error::BalanceTooLow { buy },
+            Error::LatestRateFetchFailed(_) | Error::SellQuoteCalculationFailed(_) => {
+                spot_price::Error::Other
+            }
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error<LR>
+where
+    LR: LatestRate + Debug,
```

Comment:
As a rule of thumb, try to avoid trait bounds on _type_ declarations. Traits define functionality and hence should be coupled to functionality (`impl` blocks) where possible.

There is two ways around this for this situation:

1. Use a trait object (`Box<dyn std::error::Error>`)
2. Make the `Error` generic over `E` instead of `LR::Error`, i.e.: `LatestRateFetchFailed(E)`

Both approaches will avoid cascading changes like `+ Debug` on the `LR` type parameter.